bpf/lib/srv6.h

```diff
@@ -383,18 +385,45 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_SRV6_DECAP)
 int tail_srv6_decap(struct __ctx_buff *ctx)
 {
 	int ret = 0;
+	__u8 nexthdr = 0;
+	__s8 ext_err = 0;
 
-	ret = srv6_decapsulation(ctx);
+	ret = srv6_decapsulation(ctx, &nexthdr);
 	if (ret < 0)
 		goto error_drop;
 
-	send_trace_notify(ctx, TRACE_TO_STACK, SECLABEL_IPV6, UNKNOWN_ID,
-			  TRACE_EP_ID_UNKNOWN,
-			  TRACE_IFINDEX_UNKNOWN, TRACE_REASON_SRV6_DECAP, 0);
-	return CTX_ACT_OK;
+	/* Clear all metadata before recirculation to avoid the accidental use. */
+	ctx_store_meta(ctx, 0, 0);
+	ctx_store_meta(ctx, 1, 0);
+	ctx_store_meta(ctx, 2, 0);
+	ctx_store_meta(ctx, 3, 0);
+	ctx_store_meta(ctx, 4, 0);
+
+	switch (nexthdr) {
+#ifdef ENABLE_IPV4
+	case IPPROTO_IPIP:
+		send_trace_notify(ctx, TRACE_FROM_NETWORK, SECLABEL_IPV4, UNKNOWN_ID,
+				  TRACE_EP_ID_UNKNOWN, TRACE_IFINDEX_UNKNOWN,
+				  TRACE_REASON_SRV6_DECAP, 0);
+		ret = tail_call_internal(ctx, CILIUM_CALL_IPV4_FROM_NETDEV, &ext_err);
+		return send_drop_notify_error_ext(ctx, SECLABEL_IPV6, ret, ext_err,
+						  CTX_ACT_DROP, METRIC_INGRESS);
+#endif /* ENABLE_IPV4 */
+#ifdef ENABLE_IPV6
+	case IPPROTO_IPV6:
+		send_trace_notify(ctx, TRACE_FROM_NETWORK, SECLABEL_IPV6, UNKNOWN_ID,
+				  TRACE_EP_ID_UNKNOWN, TRACE_IFINDEX_UNKNOWN,
+				  TRACE_REASON_SRV6_DECAP, 0);
+		ret = tail_call_internal(ctx, CILIUM_CALL_IPV6_FROM_NETDEV, &ext_err);
+		return send_drop_notify_error_ext(ctx, SECLABEL_IPV6, ret, ext_err,
+						  CTX_ACT_DROP, METRIC_INGRESS);
+#endif /* ENABLE_IPV6 */
```

Comment:
Ideally this would pass through a `from_srv6` flag, that gets stored into the `CT_EGRESS` entry. So that bpf_lxc can pick it up in the `ct_has_nodeport_egress_entry4()` check, and apply the RevDNAT *only* to SRv6-originating connections. Would that make sense?

Reply:
What is a purpose of introducing a new flag? In theory, it is not required for conntrack to distinguish SRv6-overlay-originated connections and others. I guess it is an alternative of `ct_state->node_port`, so that we don't need [this line](https://github.com/cilium/cilium/pull/34365/files#diff-1e6ee3d7f22bd39ab87f7d8d17b8f1c909de590f8a2f694e32a6508fad98103bR1981)?

Reply:
To reduce the blast radius, and only special-case the traffic that **really** needs it. But no big deal if it's too invasive.

Reply:
Yeah, while I agree with the idea to reduce the blast radius, adding a new field to the conntrack entry is too intrusive for other code path. We recirculate back to the CILIUM_CALL_IPV4_FROM_NETDEV here and the next conntrack hook is few functions away. That means we need to carry this `from_srv6` information with function argument / skb->cb. I personally don't want to invade to this generic code path. How about adding `&& ENABLE_SRV6` [here](https://github.com/cilium/cilium/blob/main/bpf/bpf_lxc.c#L602)?

Reply:
Fair enough :).

> How about adding `&& ENABLE_SRV6` [here](https://github.com/cilium/cilium/blob/main/bpf/bpf_lxc.c#L602)?

Eventually, yes. But we'll need to keep it enabled unconditionally for a release or two, until we can be reasonably sure that all "legacy" connections in a user's cluster have been re-established (along https://github.com/cilium/cilium/pull/32642).